{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.7.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-secondary\" style=\"font-size:30px\">\nEDA Enzyme Stability\n</div>\n\n<div class=\"alert alert-block alert-warning\" style=\"font-size:15px\">\nHi dear lector üëÄ, if this notebook is useful for you, please give us an upvote ‚ù§Ô∏èüöÄ\n</div>\n\n<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nCompetition Goals\n</div>\n\n\nThe goal of this competition is to predict the thermostability of enzyme variants. The experimentally measured thermostability (melting temperature) data includes natural sequences, as well as engineered sequences with single or multiple mutations upon the natural sequences.\n\nUnderstanding and accurately predict protein stability is a fundamental problem in biotechnology. Its applications include enzyme engineering for addressing the world‚Äôs challenges in sustainability, carbon neutrality and more. Improvements to enzyme stability could lower costs and increase the speed scientists can iterate on concepts. ([Novozymes Enzyme Stability Prediction | Kaggle, s.¬†f.](https://www.kaggle.com/competitions/novozymes-enzyme-stability-prediction))\n\n\nEnzymes can be a little complex to understand, so seeing [this video](https://www.youtube.com/watch?v=rlH1ym916Fo) will make you understand better about what are the enzymes in any case you need some visual help.","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is an enzyme and how it works\n</div>\n\n\n<img src=\"https://img.freepik.com/vector-premium/diagrama-ilustracion-vectorial-plantilla-ciclo-catalitico-enzimatico_683773-180.jpg?w=740\" >\n\n[Image Resource](https://www.freepik.es/vector-premium/diagrama-ilustracion-vectorial-plantilla-ciclo-catalitico-enzimatico_30160592.htm#query=enzyme&position=9&from_view=search&track=sph)\n\n\nEnzymes are proteins that act as catalysts in the chemical reactions of living organisms, it means that enzymes accelerate reaction speed, modifying substances called substrates, and the substrates which are chosen to bind with the enzymes to be modified, will depend in each enzyme. normally enzymes are proteins but also can be RNA.\n\n\n<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nHow enzymes speed up chemical reactions?\n</div>\n\nEnzymes accomplish this task by reducing the reaction activation energy. The amount of energy required to start a reaction varies depending on the type of enzyme, and this is where we will discuss our objective prediction, \"thermostability\".\n\nAfter enzymes binds to the substrate it makes bond breaking and bond forming, binding substrates and creating a new product as result. All this process happens into a special place called \"active site\", this place has a specific size, behavior and shape.\n\nThis shape belongs to the link between the amino acids from the enzyme, the enzymes has these compressed amino acids in one or more polypeptide chains, the size of this chains will define which substrates are \"compatible\" with the enzyme.\n\n<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nEnzyme components\n</div>\n\n<img src=\"https://www.learninsta.com/wp-content/uploads/2021/06/Enzymes-img-4.png\" >\n\n[Image Resource](https://www.learninsta.com/wp-content/uploads/2021/06/Enzymes-img-4.png)\n\n* Cations:\nPositively charged metal irons which act as activators, these cations are added temporarily to the active site to activate the enzyme (named cofactor in the image).\n* Coenzymes:\nCoenzymes are vitamin products which join to the enzyme-substrate complex temporarily!\n* Prosthetic Group:\nIt is a permanently enzyme bound\n\n\nAll this system is called Holoenzyme and an inactivated enzyme (without the other components), is called inactivated enzyme.","metadata":{}},{"cell_type":"markdown","source":"# Defining libraries, paths and constants","metadata":{}},{"cell_type":"code","source":"%%capture\n!pip install blosum\n!pip install biopandas\nimport sys\n# SYS installations\n\n\n!cp ../input/rapids/rapids.0.14.0 /opt/conda/envs/rapids.tar.gz\n!cd /opt/conda/envs/ && tar -xzvf rapids.tar.gz > /dev/null\n! wget https://ftp.ncbi.nih.gov/blast/matrices/BLOSUM100 -O BLOSUM100.txt\nsys.path = [\"/opt/conda/envs/rapids/lib/python3.7/site-packages\"] + sys.path\nsys.path = [\"/opt/conda/envs/rapids/lib/python3.7\"] + sys.path\nsys.path = [\"/opt/conda/envs/rapids/lib\"] + sys.path \n!cp /opt/conda/envs/rapids/lib/libxgboost.so /opt/conda/lib/\n\n!nvidia-smi\n\n# Check CUDA/cuDNN Version\n!nvcc -V && which nvcc","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.214749Z","iopub.status.idle":"2023-01-02T17:03:41.215145Z","shell.execute_reply.started":"2023-01-02T17:03:41.214956Z","shell.execute_reply":"2023-01-02T17:03:41.214974Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Base libraries\n\nimport numpy as np \nimport pandas as pd \nimport random\nimport requests\nfrom biopandas.pdb import PandasPdb\nfrom sklearn.decomposition import PCA\nfrom sklearn.linear_model import LinearRegression\nfrom sklearn.preprocessing import StandardScaler\nfrom scipy.stats import rankdata\nfrom tqdm.auto import tqdm\n\n# Files management\nimport os\n\n# Ploting\nimport matplotlib.pyplot as plt\nimport plotly.express as px\nimport seaborn as sns\n\n# Libraris for Data process\nimport blosum as bl\nimport Levenshtein\nfrom Levenshtein import distance as levenshtein_distance\nfrom Bio.SubsMat import MatrixInfo\n#import cudf\n","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2023-01-02T17:03:41.216141Z","iopub.status.idle":"2023-01-02T17:03:41.216538Z","shell.execute_reply.started":"2023-01-02T17:03:41.216342Z","shell.execute_reply":"2023-01-02T17:03:41.216360Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"* Cudf librarie tutoral: https://www.kaggle.com/code/beniel/rapids-cudf-tutorial\n* Levensthein Distance Explanation: https://www.youtube.com/watch?v=MiqoA-yF-0M","metadata":{}},{"cell_type":"code","source":"# Paths\n\n\n## Novozymes Database\ntestCsvPath = \"../input/novozymes-enzyme-stability-prediction/test.csv\"\ntrainCsvPath = \"../input/novozymes-enzyme-stability-prediction/train.csv\"\n\nsampleSubmissionPath = \"../input/novozymes-enzyme-stability-prediction/sample_submission.csv\"\ntrainUpdatesPath = \"../input/novozymes-enzyme-stability-prediction/train_updates_20220929.csv\"\n\npdbPath = \"/kaggle/input/novozymes-enzyme-stability-prediction/wildtype_structure_prediction_af2.pdb\"\n\n\n## External Databases\n\ntrainDFGroupPath = \"../input/train-data-contains-mutations-like-test-data/train_with_groups.csv\"","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.218759Z","iopub.status.idle":"2023-01-02T17:03:41.219527Z","shell.execute_reply.started":"2023-01-02T17:03:41.219158Z","shell.execute_reply":"2023-01-02T17:03:41.219191Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Understanding dataset","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\ntestDF\n</div>\n","metadata":{}},{"cell_type":"code","source":"testDF = pd.read_csv(testCsvPath)\ntestDF.head()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.221727Z","iopub.status.idle":"2023-01-02T17:03:41.222394Z","shell.execute_reply.started":"2023-01-02T17:03:41.222069Z","shell.execute_reply":"2023-01-02T17:03:41.222108Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDF.info","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.226023Z","iopub.status.idle":"2023-01-02T17:03:41.226672Z","shell.execute_reply.started":"2023-01-02T17:03:41.226364Z","shell.execute_reply":"2023-01-02T17:03:41.226394Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\ntrainDF\n</div>\n","metadata":{}},{"cell_type":"code","source":"trainDF = pd.read_csv(trainCsvPath)\nprint(trainDF[\"data_source\"][0])\nprint(trainDF[\"data_source\"][10])\nprint(trainDF[\"data_source\"][100])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.228831Z","iopub.status.idle":"2023-01-02T17:03:41.229444Z","shell.execute_reply.started":"2023-01-02T17:03:41.229138Z","shell.execute_reply":"2023-01-02T17:03:41.229167Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"trainDF.head()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.231666Z","iopub.status.idle":"2023-01-02T17:03:41.232434Z","shell.execute_reply.started":"2023-01-02T17:03:41.232116Z","shell.execute_reply":"2023-01-02T17:03:41.232155Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"len(trainDF[\"seq_id\"])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.233628Z","iopub.status.idle":"2023-01-02T17:03:41.234076Z","shell.execute_reply.started":"2023-01-02T17:03:41.233860Z","shell.execute_reply":"2023-01-02T17:03:41.233881Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"trainDF[\"protein_sequence\"][0]","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.235752Z","iopub.status.idle":"2023-01-02T17:03:41.236201Z","shell.execute_reply.started":"2023-01-02T17:03:41.236003Z","shell.execute_reply":"2023-01-02T17:03:41.236023Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"## As we can see, we have only 325 pages where comes from our data\nlen(pd.unique(trainDF[\"data_source\"]))","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.237975Z","iopub.status.idle":"2023-01-02T17:03:41.238334Z","shell.execute_reply.started":"2023-01-02T17:03:41.238155Z","shell.execute_reply":"2023-01-02T17:03:41.238172Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nSample Submission\n</div>\n","metadata":{}},{"cell_type":"code","source":"sampleSubmission = pd.read_csv(sampleSubmissionPath)\nsampleSubmission.head()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.240541Z","iopub.status.idle":"2023-01-02T17:03:41.241224Z","shell.execute_reply.started":"2023-01-02T17:03:41.241016Z","shell.execute_reply":"2023-01-02T17:03:41.241037Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"sampleSubmission.info","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.243251Z","iopub.status.idle":"2023-01-02T17:03:41.243754Z","shell.execute_reply.started":"2023-01-02T17:03:41.243550Z","shell.execute_reply":"2023-01-02T17:03:41.243571Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nTrain Updates\n</div>","metadata":{}},{"cell_type":"code","source":"trainUpdates = pd.read_csv(trainUpdatesPath)\ntrainUpdates","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.245448Z","iopub.status.idle":"2023-01-02T17:03:41.246142Z","shell.execute_reply.started":"2023-01-02T17:03:41.245926Z","shell.execute_reply":"2023-01-02T17:03:41.245946Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"trainUpdates.info","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.247365Z","iopub.status.idle":"2023-01-02T17:03:41.248135Z","shell.execute_reply.started":"2023-01-02T17:03:41.247868Z","shell.execute_reply":"2023-01-02T17:03:41.247902Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Important Technicalities\n\nIn the data we can see the next columns: \n\n* Protein sequence\n* pH\n* tm\n* data_source\n\nLets explain each column","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\npH\n</div>\n\nAccording to Brittanica, pH is a quantitative measure of acidity, basicity of aqueus in liquid solutions([The Editors of Encyclopaedia Britannica, 2022](https://www.britannica.com/science/pH)).\n\nSince pH has an impact on enzyme activity, it is crucial to monitor pH in relation to enzymes. However, there is a certain pH level at which enzyme activity will be at its peak; for example, see the example below.\n\n<img src=\"https://bam.files.bbci.co.uk/bam/live/content/zskcd2p/large\" >\n\n[Image Resource](https://www.bbc.co.uk/bitesize/guides/zcr74qt/revision/5#:~:text=Enzyme%20activity%20is%20at%20its,pH%20the%20enzyme%20activity%20decreases.)\n\nAmino acids present in the active site are acidic or basic and the fluctuation in pH can affect these amino acids, remember that this is so important because amino acids helps to the enzyme to define which substrate are going to be joined, and with pH fluctuations can make harder to the substrates to bind and extreme pH values can denature enzymes.","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\ntm\n</div>\n\nIn this case tm is referring to the thermostability measure, it is very important for industrial applications, measuring this have a special method, for better understanding how to measure see [this article](https://iubmb.onlinelibrary.wiley.com/doi/pdf/10.1002/bmb.21127).\nAlso this is going to be our focus goal prediction, the production as result of enzymes process will depends in a big way of this metric.\n\n<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nHow thermostability affects Enzymes cicle?\n</div>\n\nEach enzyme's active site is extremely sensitive, and even a small change in the environment can cause the enzyme to respond differently. Thermoestability is one of the elements that affects the active site and, as a result, how well the enzyme works.\n\nThe suitable temperature for enzymes to function properly is 37¬∫C, increasing or decreasing the temperature above 37¬∫C can affect chemical bonds in the active site making them less suited to bind substrates and higher temperatures denature enzymes ph. ([Effect of Temperature on Enzymatic Reaction - Creative Enzymes, s.¬†f.](https://www.creative-enzymes.com/resource/effect-of-temperature-on-enzymatic-reaction_50.html#:~:text=The%20optimum%20temperature%20for%20most,enzymes%20adapted%20to%20higher%20temperatures.))\n\n","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nProtein Sequence\n</div>\n‚Äã\n‚Äã\nProtein sequence is the method used to determinate all the amino acids in a protein (remember enzymes are conformed by proteins or ARN), it is until to identify it in a sample and categorize its post-translational modifications. The process of determining the amino acid sequence is known as protein sequencing. [(Smith, 2019)](https://www.news-medical.net/life-sciences/Amino-Acids-and-Protein-Sequences.aspx)\n‚Äã\n<img src=\"https://d2jx2rerrg6sh3.cloudfront.net/image-handler/ts/20170523082107/ri/673/picture/2017/5/DNA_translation_to_prtoein_680x_-_udaix.jpg\" >\n‚Äã\n[Image Resource](https://d2jx2rerrg6sh3.cloudfront.net/image-handler/ts/20170523082107/ri/673/picture/2017/5/DNA_translation_to_prtoein_680x_-_udaix.jpg)","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nData source\n</div>\n\nThis column is no complex to explain, each value into each row from this column is going to returned a link which say us where was take the information.","metadata":{}},{"cell_type":"markdown","source":"### Amino Acids Metrics","metadata":{}},{"cell_type":"markdown","source":"\nAccording to Evolutions and genomics, the divergence among sequences can be modeled with a mutation matrix. [(Amino acid substitution models, 2016)](https://evomics.org/resources/substitution-models/amino-acid-substitution-models/#:~:text=Empirical%20substitution%20models&text=In%20the%20Dayhoff%20approach%2C%20replacement,of%20successive%20mutations%20is%20low.).\nIn that reference we cand find that there are differents ways to see the divergences and mutations in amino acids, between them exists:\n\n* PAM Matrices\n* Dayhoff Matrices\n* JIT Matrices\n* BLOSUM\n* Poisson models\n\nBlocks Amino Acid Substitution Matrices is a substition metric used for sequences alignment in proteins, a lecture talking about this deeply is the(Gilbert et¬†al., 1992)\n\nWhat is the sequence alignment for?\n\nA sequence alignment is a way of arranging the primary sequences of a protein to identify regions of similarity that may be a consequence of functional, structural, or evolutionary relationships between the sequences.([Aligning multiple protein sequences | UniProt, s.¬†f.](https://www.ebi.ac.uk/training/online/courses/uniprot-exploring-protein-sequence-and-functional-info/how-to-use-uniprot-tools-clone/aligning-multiple-protein-sequences/#:~:text=A%20sequence%20alignment%20is%20a,evolutionary%20relationships%20between%20the%20sequences.))\n\nPlease see the next video to understand better what are we doing in the next code blocks\nhttps://www.youtube.com/watch?v=MiqoA-yF-0M","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nTest Data Mutations\n</div>\n","metadata":{}},{"cell_type":"code","source":"# Wild type sequence provided in the \"Dataset Description\":\nbase = 'VPVNPEPDATSVENVALKTGSGDSQSDPIKADLEVKGQSALPFDVDCWAILCKGAPNVLQRVNEKTKNSNRDRSGANKGPFKDPQKWGIKALPPKNPSWSAQDFKSPEEYAFASSLQGGTNAILAPVNLASQNSQGGVLNGFYSANKVAQFDPSKPQQTKGTWFQITKFTGAAGPYCKALGSNDKSVCDKNKNIAGDWGFDPAKWAYQYDEKNNKFNYVGK'","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.250192Z","iopub.status.idle":"2023-01-02T17:03:41.250563Z","shell.execute_reply.started":"2023-01-02T17:03:41.250380Z","shell.execute_reply":"2023-01-02T17:03:41.250398Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"listOutput = []","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.253264Z","iopub.status.idle":"2023-01-02T17:03:41.253701Z","shell.execute_reply.started":"2023-01-02T17:03:41.253499Z","shell.execute_reply":"2023-01-02T17:03:41.253520Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"for i in range(len(testDF)):\n    rowProteinSequence = testDF.iloc[i].protein_sequence ## Taking a row\n    LevenstreinDist = Levenshtein.editops(rowProteinSequence, base) ## Using Levenshtein\n    listOutput.append(LevenstreinDist)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.255903Z","iopub.status.idle":"2023-01-02T17:03:41.256326Z","shell.execute_reply.started":"2023-01-02T17:03:41.256130Z","shell.execute_reply":"2023-01-02T17:03:41.256149Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"def deal(s):\n    if len(s):\n        return s[0][2]\n    else:\n        return -1","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.257886Z","iopub.status.idle":"2023-01-02T17:03:41.258688Z","shell.execute_reply.started":"2023-01-02T17:03:41.258470Z","shell.execute_reply":"2023-01-02T17:03:41.258492Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testOutputs = [deal(s) for s in listOutput]\nsubScores = []","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.260615Z","iopub.status.idle":"2023-01-02T17:03:41.261057Z","shell.execute_reply.started":"2023-01-02T17:03:41.260841Z","shell.execute_reply":"2023-01-02T17:03:41.260860Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"listOutput[:5]","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.262758Z","iopub.status.idle":"2023-01-02T17:03:41.263245Z","shell.execute_reply.started":"2023-01-02T17:03:41.263047Z","shell.execute_reply":"2023-01-02T17:03:41.263066Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDF['modif'] = testOutputs\ntestDF.head()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.264596Z","iopub.status.idle":"2023-01-02T17:03:41.265521Z","shell.execute_reply.started":"2023-01-02T17:03:41.265260Z","shell.execute_reply":"2023-01-02T17:03:41.265298Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nProtein Data Bank (PDB)\n</div>\n‚Äã\n‚Äã\nProtein Data Bank (PDB) format is a standard for files containing atomic coordinates. Structures deposited in the Protein Data Bank at the Research Collaboratory for Structural Bioinformatics (RCSB) written in this standardized format https://www.wwpdb.org/. While this short description will suffice for many users, for more details, consult the definitive description link: https://www.wwpdb.org/documentation/file-format.\n\n#### Description\nProtein Data Bank format consists of lines of information in a text file. Each line of information in the file is called a record. A file generally contains several different types of records, which are arranged in a specific order to describe a structure.\n\n\n\n#### Protein Data Bank Record Types\n|Record Type||Data Provided by Record|\n|-----------||-----------------------|\n|ATOM||atomic coordinate record containing the X,Y,Z orthogonal √Ö coordinates for atoms in standard residues (amino acids and nucleic acids).|\n|HETATM||atomic coordinate record containing the X,Y,Z orthogonal √Ö coordinates for atoms in nonstandard residues. Nonstandard residues include inhibitors, cofactors, ions, and solvent. The only functional difference from ATOM records is that HETATM residues are by default not connected to other residues. Note that water residues should be in HETATM records.|\n|TER||indicates the end of a chain of residues. For example, a hemoglobin molecule consists of four subunit chains that are not connected. TER indicates the end of a chain and prevents the display of a connection to the next chain.|\n|HELIX||indicates the location and type (right-handed alpha, etc.) of helices. One record per helix.|\n|SHEET||indicates the location, sense (anti-parallel, etc.) and registration with respect to the previous strand in the sheet (if any) of each strand in the model. One record per strand.|\n|SSBOND||defines disulfide bond linkages between cysteine residues.|\n\n\nTable above describes the format for selected record types. From the standpoint of most users, the most notable differences between older and newer files occur in the fields following the temperature factor in ATOM and\nHETATM records. These fields are not included in the examples in the subsequent sections. Furthermore,\nsome fields are frequently blank, such as the alternate location indicator when an atom does not have alternate locations.\n\n\n\n\n`Copyright Ô£© 2002, The Regents of the University of California All Rights Reserved`\n‚Äã\n<img src=\"https://proteopedia.org/wiki/images/thumb/0/0c/Pdb_file_diagram.png/800px-Pdb_file_diagram.png\" >\n‚Äã\n[Image Resource](https://proteopedia.org/wiki/index.php/Image:Pdb_file_diagram.png)","metadata":{}},{"cell_type":"code","source":"pdbDF =  PandasPdb().read_pdb(pdbPath)\npdbDF.df.keys()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.267631Z","iopub.status.idle":"2023-01-02T17:03:41.268063Z","shell.execute_reply.started":"2023-01-02T17:03:41.267855Z","shell.execute_reply":"2023-01-02T17:03:41.267875Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"atomDF = pdbDF.df['ATOM']\nprint(f\"ATOM dataset is of shape: {atomDF.shape}\")","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.269456Z","iopub.status.idle":"2023-01-02T17:03:41.270417Z","shell.execute_reply.started":"2023-01-02T17:03:41.270210Z","shell.execute_reply":"2023-01-02T17:03:41.270231Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"mapNumberToB = atomDF.groupby('residue_number').b_factor.mean()\ntestDF['b_factor'] = testDF.modif.map(mapNumberToB).fillna(0)\ntestDF.head()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.271969Z","iopub.status.idle":"2023-01-02T17:03:41.272357Z","shell.execute_reply.started":"2023-01-02T17:03:41.272171Z","shell.execute_reply":"2023-01-02T17:03:41.272189Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"subMat = MatrixInfo.blosum40\nlist(subMat)[:5]","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.273651Z","iopub.status.idle":"2023-01-02T17:03:41.274064Z","shell.execute_reply.started":"2023-01-02T17:03:41.273862Z","shell.execute_reply":"2023-01-02T17:03:41.273881Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"for i in range(len(testDF)):\n    row = testDF.iloc[i]\n    edit = Levenshtein.editops(row.protein_sequence, base)\n    if len(edit) == 1:\n        if edit[0][0] == 'replace':\n            try:\n                subScore = subMat[(base[edit[0][2]], row.protein_sequence[edit[0][1]])]\n            except KeyError:\n                subScore = subMat[(row.protein_sequence[edit[0][1]], base[edit[0][2]])]\n        else:\n            subScore = -10\n    else:\n        subScore = -10\n    subScores.append(subScore)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.275536Z","iopub.status.idle":"2023-01-02T17:03:41.275956Z","shell.execute_reply.started":"2023-01-02T17:03:41.275728Z","shell.execute_reply":"2023-01-02T17:03:41.275747Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"B factor is a metric also called as debye waller factor is a metric which take importance for this situacion, see this dicussion to understand better how b factor works:\n\nhttps://www.kaggle.com/competitions/novozymes-enzyme-stability-prediction/discussion/354476","metadata":{}},{"cell_type":"code","source":"testDF['score'] = subScores\ntestDF['score_adj'] = [ 1 - (1 / (1+np.exp(-x))) for x in subScores]\ntestDF['bFactorAdj'] = testDF['b_factor'] * testDF['score_adj'] \ntestDF.head(10)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.277277Z","iopub.status.idle":"2023-01-02T17:03:41.277809Z","shell.execute_reply.started":"2023-01-02T17:03:41.277543Z","shell.execute_reply":"2023-01-02T17:03:41.277565Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDF","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.279307Z","iopub.status.idle":"2023-01-02T17:03:41.279717Z","shell.execute_reply.started":"2023-01-02T17:03:41.279513Z","shell.execute_reply":"2023-01-02T17:03:41.279532Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"def change_list(wild, mutation_list):\n\n    list_output = []\n\n    for mutation in mutation_list:\n        edits = Levenshtein.editops(wild, mutation)\n        if len(edits):\n            list_output.append(edits[0]+tuple(mutation[edits[0][1]])+tuple(base[edits[0][1]]))\n        else:\n            list_output.append(('replace', 0, 0, 'A', 'A'))\n\n    changes = pd.DataFrame(list_output,columns=['operation','position1','position2','change1','change2'])\n    changes.change2 = np.where(changes.operation=='delete','',changes.change2)\n    \n    return changes","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.281280Z","iopub.status.idle":"2023-01-02T17:03:41.281678Z","shell.execute_reply.started":"2023-01-02T17:03:41.281486Z","shell.execute_reply":"2023-01-02T17:03:41.281505Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"changes = change_list(base, testDF.protein_sequence.to_list())\nchanges.operation.value_counts()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.282907Z","iopub.status.idle":"2023-01-02T17:03:41.283296Z","shell.execute_reply.started":"2023-01-02T17:03:41.283105Z","shell.execute_reply":"2023-01-02T17:03:41.283123Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"changes","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.285250Z","iopub.status.idle":"2023-01-02T17:03:41.286003Z","shell.execute_reply.started":"2023-01-02T17:03:41.285756Z","shell.execute_reply":"2023-01-02T17:03:41.285803Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Adding changes to the testDF \ntest_df = testDF.merge(changes, left_index=True,right_index = True)\ntest_df","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.287305Z","iopub.status.idle":"2023-01-02T17:03:41.288007Z","shell.execute_reply.started":"2023-01-02T17:03:41.287780Z","shell.execute_reply":"2023-01-02T17:03:41.287808Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Saving test_df as csv\ntest_df.to_csv('test_data.csv', index=False)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.289134Z","iopub.status.idle":"2023-01-02T17:03:41.289862Z","shell.execute_reply.started":"2023-01-02T17:03:41.289625Z","shell.execute_reply":"2023-01-02T17:03:41.289646Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"pd.crosstab(changes.change1,changes.change2).style.background_gradient(axis=None, cmap=\"YlGnBu\")","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.291304Z","iopub.status.idle":"2023-01-02T17:03:41.291661Z","shell.execute_reply.started":"2023-01-02T17:03:41.291483Z","shell.execute_reply":"2023-01-02T17:03:41.291502Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nData Mutations\n</div>\n\n\nAs is metioned in this [notebook](https://www.kaggle.com/code/cdeotte/train-data-contains-mutations-like-test-data#Train-Data-Contains-Mutations), the data contains mutations, those mutations can be useful to make better predictions but, first we need to add new columns that allow us to identify groups, for that reason we are going to use the database generated by the same notebook.\n\nSome new values saved in the csv are:\n\n* column group which identifies the groups of similar proteins that are mutations\n\n\n\nWhy we have - 1 groups? Here are some words from the notebook owner:\n\nBefore using the new column group with GroupKFold, we must assign every row with group=-1 to a new unique group number (otherwise all the group=-1 will be in the same fold when they are actually different). So for example nrow = train.loc[train.group==-1].shape[0] and then mx = train.group.max() and finally train.loc[train.group==-1,'group'] = np.arange(nrow) + mx + 1.\n\n\nAlso the next code is based in this [notebook](https://www.kaggle.com/code/lucasmorin/nesp-changes-eda-and-baseline) with some modifications.","metadata":{}},{"cell_type":"code","source":"trainDFGroup = pd.read_csv(trainDFGroupPath) ## Reading data from train-data-contains-mutations\ntrainDFGroup","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.293136Z","iopub.status.idle":"2023-01-02T17:03:41.293494Z","shell.execute_reply.started":"2023-01-02T17:03:41.293316Z","shell.execute_reply":"2023-01-02T17:03:41.293332Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"## Let's see the groups in dataframe\nnp.unique(trainDFGroup[\"group\"])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.294749Z","iopub.status.idle":"2023-01-02T17:03:41.295531Z","shell.execute_reply.started":"2023-01-02T17:03:41.295318Z","shell.execute_reply":"2023-01-02T17:03:41.295338Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# remove group -1:\ntrainDF = trainDFGroup[trainDFGroup.group!=-1] ## Avoiding -1 groups\ncountByGroup = trainDF.groupby('group').seq_id.count() ## Counting proteins by group in dataframe\ncountByGroup","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.296963Z","iopub.status.idle":"2023-01-02T17:03:41.297404Z","shell.execute_reply.started":"2023-01-02T17:03:41.297192Z","shell.execute_reply":"2023-01-02T17:03:41.297211Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"def build_change_list(groupDF):\n    \n    listOutput = []\n    groupSize = len(groupDF)\n    groupValues = groupDF.values\n    \n    col = ['pH','data_source','tm','group']\n\n    for i in range(groupSize):\n        ## Defining constants\n        data1 = groupValues[i]\n        seqString1 = data1[1]\n        values1  = data1[2:]\n        \n        \n        for j in range(groupSize):\n            data2 = groupValues[j]\n            seqString2 = data2[1]\n            values2  = data2[2:]\n            \n\n            ## In this \"if\", we are taking sure to no make comparations or use editops with the same origin values\n            if i != j:\n                edits = Levenshtein.editops(seqString1, seqString2)\n                levenshteinTuple = tuple([seqString1,seqString2])\n                \n                ## Checking if edits values does exist\n                if len(edits)==1:\n                    listOutput.append(levenshteinTuple + edits[0]+ tuple(seqString1[edits[0][1]]) + tuple(seqString2[edits[0][1]]) + tuple(values1) + tuple(values2))\n                else:\n                    listOutput.append(levenshteinTuple + ('replace', 0, 0, 'A', 'A') + tuple(values1) + tuple(values2))\n                    \n\n    changes = pd.DataFrame(listOutput,columns=['protSeq1','protSeq2','operation','position1','position2','change1','change2']+[c+'1' for c in col] + [c+'2' for c in col])\n    changes.change2 = np.where(changes.operation=='delete','',changes.change2)\n    \n    return changes","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.298581Z","iopub.status.idle":"2023-01-02T17:03:41.298969Z","shell.execute_reply.started":"2023-01-02T17:03:41.298783Z","shell.execute_reply":"2023-01-02T17:03:41.298800Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nTrain Data Mutations\n</div>\n","metadata":{}},{"cell_type":"code","source":"%%time\n\ntop_n = 10 #408\nmain_groups = countByGroup.sort_values(ascending=False).index[:top_n]\n\ndf_list = []\n\nfor i in main_groups:\n    print(f'group {i}')\n    group_df = trainDF[trainDF.group==i]\n    df = build_change_list(group_df)\n    df_list.append(df)\n\ndfSm = pd.concat(df_list)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.300214Z","iopub.status.idle":"2023-01-02T17:03:41.300571Z","shell.execute_reply.started":"2023-01-02T17:03:41.300395Z","shell.execute_reply":"2023-01-02T17:03:41.300413Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"dfSm","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.302639Z","iopub.status.idle":"2023-01-02T17:03:41.303105Z","shell.execute_reply.started":"2023-01-02T17:03:41.302857Z","shell.execute_reply":"2023-01-02T17:03:41.302877Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Converting dfSm to Csv\ndfSm.to_csv('train_data.csv', index=False)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.304272Z","iopub.status.idle":"2023-01-02T17:03:41.304663Z","shell.execute_reply.started":"2023-01-02T17:03:41.304466Z","shell.execute_reply":"2023-01-02T17:03:41.304491Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# clean data a bit - same sources, change in protein, same pH and pH not too far of 7\n\ndfClean = dfSm[(dfSm.data_source1 == dfSm.data_source2)&(dfSm.pH1 == dfSm.pH2)]\ndfClean = dfClean[(dfClean.pH1>=6) & (dfClean.pH1<=8)]\ndfClean = dfClean[dfClean.position1 != 0]\n\ndfClean['target'] = dfClean['tm2'] - dfClean['tm1'] \n\nprint(len(dfClean))\ndisplay(pd.crosstab(dfClean.change1,dfClean.change2).style.background_gradient(axis=None, cmap=\"YlGnBu\"))","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.305688Z","iopub.status.idle":"2023-01-02T17:03:41.306147Z","shell.execute_reply.started":"2023-01-02T17:03:41.305916Z","shell.execute_reply":"2023-01-02T17:03:41.305935Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"dfClean.isnull().sum()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.307964Z","iopub.status.idle":"2023-01-02T17:03:41.308669Z","shell.execute_reply.started":"2023-01-02T17:03:41.308460Z","shell.execute_reply":"2023-01-02T17:03:41.308487Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Exporting dfClean to csv\ndfClean.to_csv('clean_train_data.csv', index=False)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.309953Z","iopub.status.idle":"2023-01-02T17:03:41.310605Z","shell.execute_reply.started":"2023-01-02T17:03:41.310394Z","shell.execute_reply":"2023-01-02T17:03:41.310415Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# avg target by protein changes, independtly of emplacement\n\navgTarget = dfClean.groupby(['change1','change2']).target.mean()\ndisplay(avgTarget.unstack().fillna(0).style.background_gradient(axis=None, cmap=\"RdYlBu\").format('{:.2f}'))","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.312394Z","iopub.status.idle":"2023-01-02T17:03:41.312873Z","shell.execute_reply.started":"2023-01-02T17:03:41.312646Z","shell.execute_reply":"2023-01-02T17:03:41.312666Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Extra Techniques","metadata":{}},{"cell_type":"markdown","source":"(The next part is a work in progress, in next versions we are adding references)\n\nBased in:\n[Notebook](https://www.kaggle.com/code/oxzplvifi/deletion-specific-ensemble)\n[Notebook](https://www.kaggle.com/code/awater1223/deletion-specific-ensemble-python)","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nDefining Functions and Base\n</div>\n","metadata":{}},{"cell_type":"code","source":"# Plot rank distributions\ndef plot_rank_dist(name, ax, show_del=False):\n\n    sns.kdeplot(\n        data=testDf.query('type==\"SUB\"'),\n        x='{}_rank'.format(name),\n        bw_adjust=0.3,\n        lw=3,\n        label='SUB',\n        ax=ax,\n        color='k'\n    )\n\n    ax.vlines(\n        testDf.query('type==\"DEL\"')['{}_rank'.format(name)],\n        ax.get_ylim()[0],\n        ax.get_ylim()[1],\n        lw=5,\n        label='DEL',\n        color=two_colors[0]\n    )\n\n    ax.vlines(\n        testDf.query('type==\"WT\"')['{}_rank'.format(name)],\n        ax.get_ylim()[0],\n        ax.get_ylim()[1],\n        lw=5,\n        label='WT',\n        color=two_colors[1]\n    )\n\n    if show_del:\n        sns.kdeplot(\n            data=testDf.query('type==\"DEL\"'),\n            x='{}_rank'.format(name),\n            bw_adjust=0.3,\n            lw=3,\n            label='DEL',\n            ax=ax,\n            color=two_colors[0]\n        )\n\n        ax.vlines(\n            testDf.query('type==\"DEL\"')['{}_rank'.format(name)],\n            ax.get_ylim()[0],\n            ax.get_ylim()[1],\n            lw=5,\n            label='DEL',\n            color=two_colors[0]\n        )\n\n    ax.set_xlim(-50,2550)\n    ax.set_title('{} rank distribution'.format(name), fontsize=20)\n    ax.set_xlabel('{}_rank'.format(name), fontsize=20)\n    ax.set_ylabel('Density', fontsize=20)\n\n    ax.tick_params(labelsize=12)\n    ax.legend(loc=1)\n\n    return ax","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.314335Z","iopub.status.idle":"2023-01-02T17:03:41.315016Z","shell.execute_reply.started":"2023-01-02T17:03:41.314808Z","shell.execute_reply":"2023-01-02T17:03:41.314831Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nMounting Base\n</div>\n","metadata":{}},{"cell_type":"code","source":"# Read testing set sequences and pH:\ntest_df = pd.read_csv('../input/novozymes-enzyme-stability-prediction/test.csv')\ntwo_colors = sns.xkcd_palette(['red', 'bright blue'])\n","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.316278Z","iopub.status.idle":"2023-01-02T17:03:41.316991Z","shell.execute_reply.started":"2023-01-02T17:03:41.316761Z","shell.execute_reply":"2023-01-02T17:03:41.316802Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"In the next code we are getting data manually as levenstein substitution, What is the difference with prevous method? In this we are saving which sequences does have:\n\n* SUB = Substitution\n* DEL = Deletion\n\n\nWe also are adding the next columns:\n\n* WT corresponding to \"wildType\"\n* MUT corresponding to \"Mutation\"\n* resId corresponding to residual Id which indicate us position of mutation (modif contains index)","metadata":{}},{"cell_type":"code","source":"# Add mutation information to testing set:\nresult = []\nfor _, row in test_df.iterrows():\n    ops = Levenshtein.editops(base, row['protein_sequence'])\n    assert len(ops) <= 1\n    if len(ops) > 0 and ops[0][0] == 'replace':\n        idx = ops[0][1]\n        result.append(['SUB', idx + 1, base[idx], row['protein_sequence'][idx]])\n    elif len(ops) == 0:\n        result.append(['WT', 0, '', ''])\n    elif ops[0][0] == 'insert':\n        assert False, \"Ups\"\n    elif ops[0][0] == 'delete':\n        idx = ops[0][1]\n        result.append(['DEL', idx + 1, base[idx], '_'])\n    else:\n        assert False, \"Ups\"","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.318314Z","iopub.status.idle":"2023-01-02T17:03:41.318663Z","shell.execute_reply.started":"2023-01-02T17:03:41.318489Z","shell.execute_reply":"2023-01-02T17:03:41.318506Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf = pd.concat([testDF, pd.DataFrame(data=result, columns=['type', 'resid', 'wt', 'mut'])], axis=1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.319567Z","iopub.status.idle":"2023-01-02T17:03:41.320057Z","shell.execute_reply.started":"2023-01-02T17:03:41.319805Z","shell.execute_reply":"2023-01-02T17:03:41.319836Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.321347Z","iopub.status.idle":"2023-01-02T17:03:41.321726Z","shell.execute_reply.started":"2023-01-02T17:03:41.321539Z","shell.execute_reply":"2023-01-02T17:03:41.321557Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Blosum","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is Blosum?\n</div>\n","metadata":{}},{"cell_type":"markdown","source":"To understand deeply you can read: https://www.kaggle.com/code/dschettler8845/novo-esp-eli5-performant-approaches-lb-0-451?scriptVersionId=107187300 in any case in future versions we are adding a summarize in this notebook","metadata":{}},{"cell_type":"markdown","source":"BLOSUM (Blocks of Amino Acid Substitution Matrix) is a substitution matrix used for sequence alignment of proteins. BLOSUM matrices are used to score alignments between evolutionarily divergent protein sequences. BLOSUM is based on local alignments[(Wiki)](https://wikidoc.org/index.php/BLOSUM#:~:text=BLOSUM%20(BLOcks%20of%20Amino%20Acid,is%20based%20on%20local%20alignments.).\n\nHow it is going to be utilize in this case? Simple, BLOSUM (in this case we are using BLOSUM 100) is a table which is going to return us the values depending of the mutation RNA letter","metadata":{}},{"cell_type":"markdown","source":"From here we are going to print head dataframe before and after of each technique to understand bether what is goin on.","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.322934Z","iopub.status.idle":"2023-01-02T17:03:41.323311Z","shell.execute_reply.started":"2023-01-02T17:03:41.323127Z","shell.execute_reply":"2023-01-02T17:03:41.323145Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"def blosum_apply(row):\n    if row['type'] == 'SUB':\n        return blosum.loc[row['wt'], row['mut']]\n    elif row['type'] == 'DEL':\n        return -10\n    elif row['type'] == 'WT':\n        return 0\n    else:\n        assert False, \"Ups\"","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.325451Z","iopub.status.idle":"2023-01-02T17:03:41.325868Z","shell.execute_reply.started":"2023-01-02T17:03:41.325653Z","shell.execute_reply":"2023-01-02T17:03:41.325672Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"blosum = pd.read_csv('./BLOSUM100.txt', sep='\\s+', comment='#')\ntestDf['blosum'] = testDf.apply(blosum_apply, axis=1)\ntestDf['blosum_rank'] = rankdata(testDf['blosum'])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.327613Z","iopub.status.idle":"2023-01-02T17:03:41.328007Z","shell.execute_reply.started":"2023-01-02T17:03:41.327816Z","shell.execute_reply":"2023-01-02T17:03:41.327835Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"!rm ./BLOSUM100.txt","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.329119Z","iopub.status.idle":"2023-01-02T17:03:41.329504Z","shell.execute_reply.started":"2023-01-02T17:03:41.329310Z","shell.execute_reply":"2023-01-02T17:03:41.329328Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.330734Z","iopub.status.idle":"2023-01-02T17:03:41.331152Z","shell.execute_reply.started":"2023-01-02T17:03:41.330963Z","shell.execute_reply":"2023-01-02T17:03:41.330982Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"In this case we have added blosum and blosum_rank to the dataframe, blosum corresponding to the blosum value per mutation and blosum_rank as a average rank.","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='blosum', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.332614Z","iopub.status.idle":"2023-01-02T17:03:41.333096Z","shell.execute_reply.started":"2023-01-02T17:03:41.332899Z","shell.execute_reply":"2023-01-02T17:03:41.332918Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Deletions in this case is giving us no values after using them in BLOSUM, for that reason there can be another methods wich will be more helpful with this situation, and also after seeing it, now we know that we should use more than a method to create predictions.","metadata":{}},{"cell_type":"markdown","source":"# pLDDT","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is pLDDT?\n</div>\n","metadata":{}},{"cell_type":"markdown","source":"To understand deeply you can read: https://www.kaggle.com/code/cdeotte/difference-features-lb-0-600 in any case in future versions we are adding a summarize in this notebook","metadata":{}},{"cell_type":"markdown","source":"What is a pLDDT score?\nThe predicted local distance difference test (pLDDT) score (0-100) is a per-residue confidence score, with values greater than 90 indicating high confidence, and values below 50 indicating low confidence.([AlphaFold Error Estimates](https://www.rbvi.ucsf.edu/chimerax/data/pae-apr2022/pae.html))\n\nVideo explaining pLDDT visualization: https://www.youtube.com/watch?v=oxblwn0_PMM","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.334322Z","iopub.status.idle":"2023-01-02T17:03:41.334707Z","shell.execute_reply.started":"2023-01-02T17:03:41.334517Z","shell.execute_reply":"2023-01-02T17:03:41.334536Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Read AlphaFold2 result for wild type sequence:\nplddt = (\n    pd.read_csv('../input/novozymes-enzyme-stability-prediction/wildtype_structure_prediction_af2.pdb', sep='\\s+', header=None)[[0,5,10]]\n    .rename(columns={0:'atom', 5:'resid', 10:'plddt'})\n    .query('atom==\"ATOM\"')\n    .drop_duplicates()\n)\n\n# Add B factor to the testing set:\ntestDf = pd.merge(\n    testDf,\n    plddt,\n    left_on='resid',\n    right_on='resid',\n    how='left'\n)\n\ntestDf['plddt_rank'] = rankdata(-1*testDf['plddt'])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.336718Z","iopub.status.idle":"2023-01-02T17:03:41.337632Z","shell.execute_reply.started":"2023-01-02T17:03:41.337342Z","shell.execute_reply":"2023-01-02T17:03:41.337362Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.338980Z","iopub.status.idle":"2023-01-02T17:03:41.339628Z","shell.execute_reply.started":"2023-01-02T17:03:41.339419Z","shell.execute_reply":"2023-01-02T17:03:41.339440Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Now we have extra data corresponding to these columns:\n* Atom\n* plddt\n* plddt_rank","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='plddt', ax=ax, show_del=True)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.340970Z","iopub.status.idle":"2023-01-02T17:03:41.341349Z","shell.execute_reply.started":"2023-01-02T17:03:41.341165Z","shell.execute_reply":"2023-01-02T17:03:41.341183Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"In this case plddt is giving us more information than Blosum100 and now DEl get better values, but we should continue exploring for extra methods.","metadata":{}},{"cell_type":"markdown","source":"## Differential pLDDT","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is pLDDT?\n</div>\n","metadata":{}},{"cell_type":"markdown","source":"To understand deeply you can read: https://www.kaggle.com/code/cdeotte/difference-features-lb-0-6000 in any case in future versions we are adding a summarize in this notebook","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.342284Z","iopub.status.idle":"2023-01-02T17:03:41.343090Z","shell.execute_reply.started":"2023-01-02T17:03:41.342882Z","shell.execute_reply":"2023-01-02T17:03:41.342903Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"plddtdiff = []\n\n# Wild type result:\nwt_plddt = (\n    pd.read_csv('../input/nesp-kvigly-test-mutation-pdbs/WT_unrelaxed_rank_1_model_3.pdb', sep='\\s+')\n    .loc['ATOM'].reset_index()\n    .loc[:, ['level_4', 'MODEL']].drop_duplicates()\n    .rename(columns={'level_4':'resid', 'MODEL':'plddt'})\n    .astype({'resid':int})\n    .set_index('resid')\n)\n","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.344309Z","iopub.status.idle":"2023-01-02T17:03:41.344721Z","shell.execute_reply.started":"2023-01-02T17:03:41.344520Z","shell.execute_reply":"2023-01-02T17:03:41.344539Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"wt  = pd.read_csv('../input/nesp-kvigly-test-mutation-pdbs/A101E_unrelaxed_rank_1_model_3.pdb')","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.346624Z","iopub.status.idle":"2023-01-02T17:03:41.347079Z","shell.execute_reply.started":"2023-01-02T17:03:41.346862Z","shell.execute_reply":"2023-01-02T17:03:41.346886Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"wt.head()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.349113Z","iopub.status.idle":"2023-01-02T17:03:41.349518Z","shell.execute_reply.started":"2023-01-02T17:03:41.349321Z","shell.execute_reply":"2023-01-02T17:03:41.349339Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"wt = 'VPVNPEPDATSVENVALKTGSGDSQSDPIKADLEVKGQSALPFDVDCWAILCKGAPNVLQRVNEKTKNSNRDRSGANKGPFKDPQKWGIKALPPKNPSWSAQDFKSPEEYAFASSLQGGTNAILAPVNLASQNSQGGVLNGFYSANKVAQFDPSKPQQTKGTWFQITKFTGAAGPYCKALGSNDKSVCDKNKNIAGDWGFDPAKWAYQYDEKNNKFNYVGK'","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.350989Z","iopub.status.idle":"2023-01-02T17:03:41.351549Z","shell.execute_reply.started":"2023-01-02T17:03:41.351194Z","shell.execute_reply":"2023-01-02T17:03:41.351212Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Add difference in pLDDTto the testing set:>\nfor _,row in tqdm(testDf.iterrows(), total=testDf.shape[0]):\n    \n    file_path = '../input/nesp-kvigly-test-mutation-pdbs/{}{}{}_unrelaxed_rank_1_model_3.pdb'.format(row['wt'], row['resid'], row['mut'])\n    if os.path.exists(file_path):\n        tdf = (\n            pd.read_csv(file_path, sep='\\s+')\n            .loc['ATOM'].reset_index()\n            .loc[:, ['level_4', 'MODEL']].drop_duplicates()\n            .rename(columns={'level_4':'resid', 'MODEL':'plddt'})\n            .astype({'resid':int})\n            .set_index('resid')\n        )\n        plddtdiff.append((tdf.loc[row['resid']] - wt_plddt.loc[row['resid']]).values[0])\n    else:\n        plddtdiff.append(np.nan)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.352561Z","iopub.status.idle":"2023-01-02T17:03:41.353533Z","shell.execute_reply.started":"2023-01-02T17:03:41.353249Z","shell.execute_reply":"2023-01-02T17:03:41.353273Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf['plddtdiff'] = plddtdiff\ntestDf['plddtdiff_rank'] = rankdata(testDf['plddtdiff'])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.354560Z","iopub.status.idle":"2023-01-02T17:03:41.354962Z","shell.execute_reply.started":"2023-01-02T17:03:41.354751Z","shell.execute_reply":"2023-01-02T17:03:41.354783Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.356133Z","iopub.status.idle":"2023-01-02T17:03:41.356513Z","shell.execute_reply.started":"2023-01-02T17:03:41.356320Z","shell.execute_reply":"2023-01-02T17:03:41.356337Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Added Columns:\n* plddtdiff\n* plddtdiff_rank","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='plddtdiff', ax=ax, show_del=True)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.358530Z","iopub.status.idle":"2023-01-02T17:03:41.358939Z","shell.execute_reply.started":"2023-01-02T17:03:41.358738Z","shell.execute_reply":"2023-01-02T17:03:41.358756Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# DeepDDG","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is DeepDDG?\n</div>\n","metadata":{}},{"cell_type":"markdown","source":"To understand deeply you can read: https://www.kaggle.com/code/dschettler8845/novo-esp-eli5-performant-approaches-lb-0-451?scriptVersionId=107187300 in any case in future versions we are adding a summarize in this notebook","metadata":{}},{"cell_type":"markdown","source":"Accurately predicting changes in protein stability due to mutations is important for protein engineering and for understanding the functional consequences of missense mutations in proteins. We have developed DeepDDG, a neural network-based method, for use in the prediction of changes in the stability of proteins due to point mutations. The neural network was trained on more than 5700 manually curated experimental data points and was able to obtain a Pearson correlation coefficient of 0.48‚Äì0.56 for three independent test sets, which outperformed 11 other methods. Detailed analysis of the input features shows that the solvent accessible surface area of the mutated residue is the most important feature, which suggests that the buried hydrophobic area is the major determinant of protein stability. We expect this method to be useful for large-scale design and engineering of protein stability. ([DeepDDG, Huali Chao](https://pubs.acs.org/doi/10.1021/acs.jcim.8b00697))","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.360035Z","iopub.status.idle":"2023-01-02T17:03:41.360403Z","shell.execute_reply.started":"2023-01-02T17:03:41.360221Z","shell.execute_reply":"2023-01-02T17:03:41.360238Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Run DeepDDG on http://protein.org.cn/ddg.html by uploading the PDB file and clicking \"Submit\":>\nddg = pd.read_csv('../input/novozymes/ddgout.txt', sep='\\s+', usecols=[0,1,2,3,4]).rename(columns={'WT':'wt', 'ResID':'resid', 'Mut':'mut'})\n\n# Add DeepDDG output to the testing set:\ntestDf = pd.merge(\n    testDf.set_index(['wt','resid','mut']),\n    ddg.set_index(['wt','resid','mut']),\n    left_index=True,\n    right_index=True,\n    how='left'\n).reset_index()\n\ntestDf.loc[testDf['type']=='WT','ddG'] = 0\ntestDf.loc[testDf['type']=='DEL','ddG'] = testDf['ddG'].dropna().median()\n\ntestDf['ddG_rank'] = rankdata(testDf['ddG'])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.361683Z","iopub.status.idle":"2023-01-02T17:03:41.362097Z","shell.execute_reply.started":"2023-01-02T17:03:41.361904Z","shell.execute_reply":"2023-01-02T17:03:41.361924Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.363447Z","iopub.status.idle":"2023-01-02T17:03:41.363882Z","shell.execute_reply.started":"2023-01-02T17:03:41.363639Z","shell.execute_reply":"2023-01-02T17:03:41.363657Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Added Columns:\n* #chain\n* ddG\n* ddG_rank","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='ddG', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.365041Z","iopub.status.idle":"2023-01-02T17:03:41.365425Z","shell.execute_reply.started":"2023-01-02T17:03:41.365235Z","shell.execute_reply":"2023-01-02T17:03:41.365253Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# DeMaSk","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is DeMaSK?\n</div>","metadata":{}},{"cell_type":"markdown","source":"To understand deeply you can read: https://www.kaggle.com/code/dschettler8845/novo-esp-eli5-performant-approaches-lb-0-451?scriptVersionId=107187300 in any case in future versions we are adding a summarize in this notebook","metadata":{}},{"cell_type":"markdown","source":"DeMaSk uses a simple linear model to predict the functional impact of single amino acid substitutions. It uses an interpretable, directional amino acid substitution matrix computed from deep mutational scanning datasets, as well as per-position evolutionary conservation and variant frequency computed from homologs of the query sequence","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.368760Z","iopub.status.idle":"2023-01-02T17:03:41.369749Z","shell.execute_reply.started":"2023-01-02T17:03:41.369495Z","shell.execute_reply":"2023-01-02T17:03:41.369524Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Run DeMaSk on https://demask.princeton.edu/query/ by pasting the wild type sequence and clicking \"Compute\":\ndemask = pd.read_csv('../input/novozymes/demaskout.txt', sep='\\t', usecols=[0,1,2,3], names=['resid','wt','mut','demask'], skiprows=1)\n\n# Add DeMask output to the testing set:\ntestDf = pd.merge(\n    testDf.set_index(['wt','resid','mut']),\n    demask.set_index(['wt','resid','mut']),\n    left_index=True,\n    right_index=True,\n    how='left'\n).reset_index()\n\ntestDf.loc[testDf['type']=='WT','demask'] = 0\ntestDf.loc[testDf['type']=='DEL','demask'] = testDf['demask'].dropna().min()\n\n\ntestDf['demask_rank'] = rankdata(testDf['demask'])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.371076Z","iopub.status.idle":"2023-01-02T17:03:41.371846Z","shell.execute_reply.started":"2023-01-02T17:03:41.371605Z","shell.execute_reply":"2023-01-02T17:03:41.371626Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.373353Z","iopub.status.idle":"2023-01-02T17:03:41.374085Z","shell.execute_reply.started":"2023-01-02T17:03:41.373862Z","shell.execute_reply":"2023-01-02T17:03:41.373888Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Added Columns:\n* demask\n* demask_rank","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='demask', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.375704Z","iopub.status.idle":"2023-01-02T17:03:41.376414Z","shell.execute_reply.started":"2023-01-02T17:03:41.376195Z","shell.execute_reply":"2023-01-02T17:03:41.376216Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"#  RMSD\n\n<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is DeMaSK?\n</div>\n\nThe rmsd value gives the average deviation between the corresponding atoms of two proteins: the smaller the rmsd, the more similar the two structures. Efficient algorithms have been developed to find the best orientation of two structures that gives the minimal possible rmsd [2], [3] [scienceDirec](https://www.sciencedirect.com/science/article/pii/S1359027898000194).","metadata":{}},{"cell_type":"markdown","source":"To understand deeply you can read: https://www.kaggle.com/code/oxzplvifi/rmsd-from-molecular-dynamics in any case in future versions we are adding a summarize in this notebook\n\n","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.377668Z","iopub.status.idle":"2023-01-02T17:03:41.378364Z","shell.execute_reply.started":"2023-01-02T17:03:41.378151Z","shell.execute_reply":"2023-01-02T17:03:41.378173Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Read VMD/NAMD output:\nnamd = pd.read_csv('../input/novozymes-md2/residue_rmsd_sasa_last.dat', sep='\\t', header=None, names=['resid','rmsd','sasa0','sasaf'])\n\n# Add VMD/NAMD results to the testing set:\ntestDf = pd.merge(\n    testDf,\n    namd[['resid','rmsd']],\n    left_on='resid',\n    right_on='resid',\n    how='left'\n)\n\ntestDf.loc[testDf['type']=='WT','rmsd'] = testDf['rmsd'].dropna().max()\n# test_df.loc[test_df['type']=='WT','sasaf'] = test_df['sasaf'].dropna().max()\n\ntestDf['rmsd_rank'] = rankdata(testDf['rmsd'])\n# test_df['sasaf_rank'] = rankdata(test_df['sasaf'])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.379705Z","iopub.status.idle":"2023-01-02T17:03:41.380355Z","shell.execute_reply.started":"2023-01-02T17:03:41.380128Z","shell.execute_reply":"2023-01-02T17:03:41.380149Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.382087Z","iopub.status.idle":"2023-01-02T17:03:41.382452Z","shell.execute_reply.started":"2023-01-02T17:03:41.382269Z","shell.execute_reply":"2023-01-02T17:03:41.382286Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Added Columns:\n* rmsd\n* rmsd_rank","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='rmsd', ax=ax, show_del=True)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.383167Z","iopub.status.idle":"2023-01-02T17:03:41.383565Z","shell.execute_reply.started":"2023-01-02T17:03:41.383373Z","shell.execute_reply":"2023-01-02T17:03:41.383393Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# SASA\n\n<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is DeMaSK?\n</div>\n\nSolvent accessible surface area (SASA) of proteins has always been considered as a decisive factor in protein folding and stability studies. It is defined as the surface characterized around a protein by a hypothetical centre of a solvent sphere with the van der Waals contact surface of the molecule ([S. Ausaf](https://pubmed.ncbi.nlm.nih.gov/24678666/#:~:text=Solvent%20accessible%20surface%20area%20(SASA,contact%20surface%20of%20the%20molecule.)).\n\nTo understand deeply you can read: https://www.kaggle.com/code/oxzplvifi/rmsd-from-molecular-dynamics in any case in future versions we are adding a summarize in this notebook","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.384622Z","iopub.status.idle":"2023-01-02T17:03:41.385035Z","shell.execute_reply.started":"2023-01-02T17:03:41.384839Z","shell.execute_reply":"2023-01-02T17:03:41.384858Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"namd","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.386985Z","iopub.status.idle":"2023-01-02T17:03:41.387413Z","shell.execute_reply.started":"2023-01-02T17:03:41.387196Z","shell.execute_reply":"2023-01-02T17:03:41.387213Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Read VMD/NAMD output:\nnamd = pd.read_csv('../input/novozymes-md/residue_rmsd_sasa_last.dat', sep='\\t', header=None, names=['resid','rmsd','sasa0','sasaf'])\n\n# Add VMD/NAMD results to the testing set:\ntestDf = pd.merge(\n    testDf,\n    namd[['resid','sasaf']],\n    left_on='resid',\n    right_on='resid',\n    how='left'\n)\n\n# test_df.loc[test_df['type']=='WT','rmsd'] = test_df['rmsd'].dropna().max()\ntestDf.loc[testDf['type']=='WT','sasaf'] = testDf['sasaf'].dropna().max()\n\n# test_df['rmsd_rank'] = rankdata(test_df['rmsd'])\ntestDf['sasaf_rank'] = rankdata(testDf['sasaf'])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.390874Z","iopub.status.idle":"2023-01-02T17:03:41.391796Z","shell.execute_reply.started":"2023-01-02T17:03:41.391426Z","shell.execute_reply":"2023-01-02T17:03:41.391459Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.394462Z","iopub.status.idle":"2023-01-02T17:03:41.395134Z","shell.execute_reply.started":"2023-01-02T17:03:41.394808Z","shell.execute_reply":"2023-01-02T17:03:41.394840Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Added Columns:\n* sasaf\n* sasaf_rank","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='sasaf', ax=ax, show_del=True)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.397365Z","iopub.status.idle":"2023-01-02T17:03:41.398086Z","shell.execute_reply.started":"2023-01-02T17:03:41.397749Z","shell.execute_reply":"2023-01-02T17:03:41.397801Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Rosetta\n\n<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is DeMaSK?\n</div>\n\nThe Rosetta software suite includes algorithms for computational modeling and analysis of protein structures. It has enabled notable scientific advances in computational biology, including de novo protein design, enzyme design, ligand docking, and structure prediction of biological macromolecules and macromolecular complexes.([ROSETTA](https://www.rosettacommons.org/software))\n\nTo understand deeply you can read: https://www.kaggle.com/code/shlomoron/nesp-relaxed-rosetta-scores in any case in future versions we are adding a summarize in this notebook","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.400234Z","iopub.status.idle":"2023-01-02T17:03:41.400878Z","shell.execute_reply.started":"2023-01-02T17:03:41.400549Z","shell.execute_reply":"2023-01-02T17:03:41.400580Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf['rosetta_rank'] = pd.read_csv('../input/nesp-relaxed-rosetta-scores/submission_rosetta_scores')['tm']","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.402570Z","iopub.status.idle":"2023-01-02T17:03:41.403198Z","shell.execute_reply.started":"2023-01-02T17:03:41.402891Z","shell.execute_reply":"2023-01-02T17:03:41.402919Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.405273Z","iopub.status.idle":"2023-01-02T17:03:41.405886Z","shell.execute_reply.started":"2023-01-02T17:03:41.405563Z","shell.execute_reply":"2023-01-02T17:03:41.405592Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Added Columns:\n\n* rosseta_rank","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='rosetta', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.408399Z","iopub.status.idle":"2023-01-02T17:03:41.409013Z","shell.execute_reply.started":"2023-01-02T17:03:41.408688Z","shell.execute_reply":"2023-01-02T17:03:41.408716Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Thermonet","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-info\" style=\"font-size:15px\">\nWhat is DeMaSK?\n</div>","metadata":{}},{"cell_type":"markdown","source":"ThermoNet is a computational framework for quantitative prediction of the changes in protein thermodynamic stability () caused by single-point amino acid substitutions. The core algorithm of ThermoNet is an ensemble of deep 3D convolutional neural networks([reference](https://bio.tools/thermonet-protein#:~:text=ThermoNet%20(biotools%3Athermonet%2Dprotein)&text=ThermoNet%20is%20a%20computational%20framework,deep%203D%20convolutional%20neural%20networks.)).\n\nTo understand deeply you can read: https://www.kaggle.com/code/vslaykovsky/nesp-thermonet-v2 in any case in future versions we are adding a summarize in this notebook","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.419796Z","iopub.status.idle":"2023-01-02T17:03:41.420552Z","shell.execute_reply.started":"2023-01-02T17:03:41.420195Z","shell.execute_reply":"2023-01-02T17:03:41.420226Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf['thermonet'] = pd.read_csv('../input/nesp-thermonet-v2/submission.csv')['tm']\ntestDf['thermonet_rank'] = rankdata(testDf['thermonet'])","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.423059Z","iopub.status.idle":"2023-01-02T17:03:41.423699Z","shell.execute_reply.started":"2023-01-02T17:03:41.423386Z","shell.execute_reply":"2023-01-02T17:03:41.423418Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.425866Z","iopub.status.idle":"2023-01-02T17:03:41.426498Z","shell.execute_reply.started":"2023-01-02T17:03:41.426179Z","shell.execute_reply":"2023-01-02T17:03:41.426209Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Added Columns:\n\n* thermonet\n* thermonet_rank","metadata":{}},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='thermonet', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.428541Z","iopub.status.idle":"2023-01-02T17:03:41.429161Z","shell.execute_reply.started":"2023-01-02T17:03:41.428854Z","shell.execute_reply":"2023-01-02T17:03:41.428883Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# 3D Properties","metadata":{}},{"cell_type":"code","source":"testDf['p3'] = pd.read_csv('/kaggle/input/nesp-submissions/submission_P03_0482.csv')['tm']\ntestDf['p3_rank'] = rankdata(testDf['p3'])\n\nfig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='p3', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.430962Z","iopub.status.idle":"2023-01-02T17:03:41.431573Z","shell.execute_reply.started":"2023-01-02T17:03:41.431249Z","shell.execute_reply":"2023-01-02T17:03:41.431292Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf['p6'] = pd.read_csv('/kaggle/input/nesp-submissions/submission_P06_0460.csv')['tm']\ntestDf['p6_rank'] = rankdata(testDf['p6'])\n\nfig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='p6', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.433543Z","iopub.status.idle":"2023-01-02T17:03:41.434150Z","shell.execute_reply.started":"2023-01-02T17:03:41.433854Z","shell.execute_reply":"2023-01-02T17:03:41.433882Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf['p2'] = pd.read_csv('/kaggle/input/nesp-submissions/submission_P02_0448.csv')['tm']\ntestDf['p2_rank'] = rankdata(testDf['p2'])\n\nfig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='p2', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.436471Z","iopub.status.idle":"2023-01-02T17:03:41.437069Z","shell.execute_reply.started":"2023-01-02T17:03:41.436747Z","shell.execute_reply":"2023-01-02T17:03:41.436796Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"testDf['3D_avg_tn'] = pd.read_csv('/kaggle/input/nesp-submissions/submission_TN_0488.csv')['tm']\ntestDf['3D_avg_tn_rank'] = rankdata(testDf['3D_avg_tn'])\n\nfig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='3D_avg_tn', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2022-12-18T21:34:29.178556Z","iopub.execute_input":"2022-12-18T21:34:29.179498Z","iopub.status.idle":"2022-12-18T21:34:29.491197Z","shell.execute_reply.started":"2022-12-18T21:34:29.179451Z","shell.execute_reply":"2022-12-18T21:34:29.489916Z"}}},{"cell_type":"markdown","source":"> # RASP","metadata":{}},{"cell_type":"code","source":"testDf['rasp'] = pd.read_csv('/kaggle/input/nesp-submissions/submission_rasp_0437.csv')['tm']\ntestDf['rasp_rank'] = rankdata(testDf['rasp'])\n\nfig, ax = plt.subplots(figsize=(25, 5))\nplot_rank_dist(name='rasp', ax=ax, show_del=False)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.439504Z","iopub.status.idle":"2023-01-02T17:03:41.440133Z","shell.execute_reply.started":"2023-01-02T17:03:41.439826Z","shell.execute_reply":"2023-01-02T17:03:41.439856Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Ensembling","metadata":{}},{"cell_type":"markdown","source":"As we have seen in previous techniques each of them have different accuracys for each mutation case, so for that same reason in the next code is going to show us an algorithm wich mixes all thoses previous values.","metadata":{}},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.442274Z","iopub.status.idle":"2023-01-02T17:03:41.442896Z","shell.execute_reply.started":"2023-01-02T17:03:41.442577Z","shell.execute_reply":"2023-01-02T17:03:41.442606Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"def rank_nrom(name):\n    # Name is a value corresponding to each previous column added so yes, its is beeing saved into \"s\"\n    s = testDf['{}_rank'.format(name)]\n    # Here is beeing returned a percentage result of the operation\n    return s/s.max()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.445159Z","iopub.status.idle":"2023-01-02T17:03:41.445761Z","shell.execute_reply.started":"2023-01-02T17:03:41.445463Z","shell.execute_reply":"2023-01-02T17:03:41.445492Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Scale ranks prior to ensembling:\n# Global ensemble:\ntestDf['tm'] = (\n    4 * rank_nrom('rosetta') + 2*rank_nrom('rmsd') + 2*rank_nrom('thermonet') + 2*rank_nrom('plddtdiff') +\\\n    rank_nrom('sasaf') + rank_nrom('plddt') + rank_nrom('demask') + rank_nrom('ddG') + rank_nrom('blosum')\n) / 14\n","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.447504Z","iopub.status.idle":"2023-01-02T17:03:41.448127Z","shell.execute_reply.started":"2023-01-02T17:03:41.447820Z","shell.execute_reply":"2023-01-02T17:03:41.447849Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Scale ranks prior to ensembling:\n# Global ensemble:\ntestDf['tm'] = (\n    4 * rank_nrom('rosetta') + 2*rank_nrom('rmsd') + 2*rank_nrom('thermonet') + 2*rank_nrom('plddtdiff') +\\\n    rank_nrom('sasaf') + rank_nrom('plddt') + rank_nrom('demask') + rank_nrom('ddG') + rank_nrom('blosum') +\\\n    rank_nrom('rasp') + 0*rank_nrom('3D_avg_p3')\n) / 16\n","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.450396Z","iopub.status.idle":"2023-01-02T17:03:41.451209Z","shell.execute_reply.started":"2023-01-02T17:03:41.450692Z","shell.execute_reply":"2023-01-02T17:03:41.450721Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"submits = ['rosetta', 'rmsd', 'thermonet', 'plddtdiff', 'sasaf', 'plddt', 'demask', 'ddG', 'blosum', 'rasp', 'p3', 'p2', 'p6']\nprint('\\t', end='')\nfor s1 in submits:\n    print(s1[:5],'\\t', end='')\nprint('')\nfor s1 in submits:\n    print(s1[:5],'\\t',end='')\n    for s2 in submits:\n        tm1 = testDf[f'{s1}_rank']\n        tm2 = testDf[f'{s2}_rank']\n        print(f\"{spearmanr(tm1,tm2).correlation:.4f}\\t\",end = '')\n    print('')\n        \n        \n        \n        \n","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.453387Z","iopub.status.idle":"2023-01-02T17:03:41.453889Z","shell.execute_reply.started":"2023-01-02T17:03:41.453676Z","shell.execute_reply":"2023-01-02T17:03:41.453696Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Scale ranks prior to ensembling:\n# Global ensemble:\n#testDf['tm'] = (\n#    4 * rank_nrom('rosetta') + 2*rank_nrom('rmsd') + 2*rank_nrom('thermonet') + 2*rank_nrom('plddtdiff') +\\\n#    rank_nrom('sasaf') + rank_nrom('plddt') + rank_nrom('demask') + rank_nrom('ddG') + rank_nrom('blosum') +\\        \n#    2.5*rank_nrom('3D_avg_p3')\n#) / 14\n","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.455565Z","iopub.status.idle":"2023-01-02T17:03:41.456245Z","shell.execute_reply.started":"2023-01-02T17:03:41.455914Z","shell.execute_reply":"2023-01-02T17:03:41.455944Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Scale ranks prior to ensembling:\n# Global ensemble:\n#testDf['tm'] = (\n#    4 * rank_nrom('rosetta') + 2*rank_nrom('rmsd') + 2*rank_nrom('thermonet') + 2*rank_nrom('plddtdiff') +\\\n#    rank_nrom('sasaf') + rank_nrom('plddt') + rank_nrom('demask') + rank_nrom('ddG') + rank_nrom('blosum') +\\\n#    2*rank_nrom('3D_avg_p2')\n#) / 14","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.458982Z","iopub.status.idle":"2023-01-02T17:03:41.459624Z","shell.execute_reply.started":"2023-01-02T17:03:41.459312Z","shell.execute_reply":"2023-01-02T17:03:41.459344Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.461744Z","iopub.status.idle":"2023-01-02T17:03:41.462398Z","shell.execute_reply.started":"2023-01-02T17:03:41.462074Z","shell.execute_reply":"2023-01-02T17:03:41.462105Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Dividing each row values per max in column\ntestDf['tm'] = testDf['tm']/testDf['tm'].max()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.464484Z","iopub.status.idle":"2023-01-02T17:03:41.465138Z","shell.execute_reply.started":"2023-01-02T17:03:41.464820Z","shell.execute_reply":"2023-01-02T17:03:41.464851Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.466648Z","iopub.status.idle":"2023-01-02T17:03:41.467297Z","shell.execute_reply.started":"2023-01-02T17:03:41.466979Z","shell.execute_reply":"2023-01-02T17:03:41.467008Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Deletion type:\nidx = testDf[testDf['type']==\"DEL\"].index\ntestDf.loc[idx, 'tm'] =  (2*rank_nrom('plddt')[idx] + 2*rank_nrom('plddtdiff')[idx] + rank_nrom('rmsd')[idx] + rank_nrom('sasaf')[idx]) / 6","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.469904Z","iopub.status.idle":"2023-01-02T17:03:41.470573Z","shell.execute_reply.started":"2023-01-02T17:03:41.470236Z","shell.execute_reply":"2023-01-02T17:03:41.470276Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Wild type:\ntestDf.loc[testDf['type']==\"WT\",'tm'] = testDf['tm'].max()+1\ntestDf['tm'] = rankdata(testDf['tm'])\ntestDf['tm_rank'] = testDf['tm']","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.472493Z","iopub.status.idle":"2023-01-02T17:03:41.473316Z","shell.execute_reply.started":"2023-01-02T17:03:41.472906Z","shell.execute_reply":"2023-01-02T17:03:41.472958Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Submission:\ntestDf[['seq_id','tm']].to_csv('submission.csv', index=False)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.476156Z","iopub.status.idle":"2023-01-02T17:03:41.476596Z","shell.execute_reply.started":"2023-01-02T17:03:41.476395Z","shell.execute_reply":"2023-01-02T17:03:41.476414Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"from scipy.stats import spearmanr\n\ndf_0613 = pd.read_csv('/kaggle/input/nesp-submissions/submission_0613.csv')\ndf_0482 = pd.read_csv('/kaggle/input/nesp-submissions/submission_P03_0482.csv')\ndf_0448 = pd.read_csv('/kaggle/input/nesp-submissions/submission_P02_0448.csv')\n\nprint('sp with 0.613: ', spearmanr(testDf.tm, df_0613.tm))","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.477983Z","iopub.status.idle":"2023-01-02T17:03:41.478359Z","shell.execute_reply.started":"2023-01-02T17:03:41.478178Z","shell.execute_reply":"2023-01-02T17:03:41.478195Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"diffs = rankdata(df_0448.tm[50:80]) - rankdata(df_0482.tm[50:80])\n\nplt.plot(diffs)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.479625Z","iopub.status.idle":"2023-01-02T17:03:41.480021Z","shell.execute_reply.started":"2023-01-02T17:03:41.479830Z","shell.execute_reply":"2023-01-02T17:03:41.479848Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"diffs = rankdata(df_0482.tm[:100]) #- rankdata(df_0482.tm)\n\nplt.plot(diffs)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.481370Z","iopub.status.idle":"2023-01-02T17:03:41.481802Z","shell.execute_reply.started":"2023-01-02T17:03:41.481596Z","shell.execute_reply":"2023-01-02T17:03:41.481614Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"testDf.head(1)","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.483131Z","iopub.status.idle":"2023-01-02T17:03:41.483501Z","shell.execute_reply.started":"2023-01-02T17:03:41.483313Z","shell.execute_reply":"2023-01-02T17:03:41.483329Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"How do you select coefficients for each method in the global ensemble? Based on what?\n\nThank you, good question, my understanding from reading discussions is that the cv and lb correlation is not very stable, so I simply added +1 to the coefficients of each method until the public score stopped improving. Considering that the public score uses 50% of the test data and that 100% of the test data comes from a single protein, I am hoping this will not cause a significant overfitting to the public half of the test data; although it might still be advisable if possible to try optimizing the coefficents on both the cv and the lb separately for each of the two submissions allowed. I have not attempted yet applying all of the methods to the full training set; but this seems like a promising next step([OscarVillaReal](https://www.kaggle.com/oxzplvifi)).","metadata":{}},{"cell_type":"code","source":"    \nplot_rank_dist(name='tm', ax=ax, show_del=True)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.485010Z","iopub.status.idle":"2023-01-02T17:03:41.485416Z","shell.execute_reply.started":"2023-01-02T17:03:41.485215Z","shell.execute_reply":"2023-01-02T17:03:41.485233Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-success\" style=\"font-size:13px\">\nPloting\n</div>\n","metadata":{}},{"cell_type":"code","source":"fig, axs = plt.subplots(nrows=10, figsize=(20,40), gridspec_kw={'hspace':0.5})\n\nplot_rank_dist(name='rosetta', ax=axs[0], show_del=False)\nplot_rank_dist(name='rmsd', ax=axs[1], show_del=True)\nplot_rank_dist(name='thermonet', ax=axs[2], show_del=False)\nplot_rank_dist(name='plddtdiff', ax=axs[3], show_del=True)\nplot_rank_dist(name='sasaf', ax=axs[4], show_del=True)\nplot_rank_dist(name='plddt', ax=axs[5], show_del=True)\nplot_rank_dist(name='demask', ax=axs[6], show_del=False)\nplot_rank_dist(name='ddG', ax=axs[7], show_del=False)\nplot_rank_dist(name='blosum', ax=axs[8], show_del=False)\n\nplot_rank_dist(name='tm', ax=axs[9], show_del=True)\n\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.486907Z","iopub.status.idle":"2023-01-02T17:03:41.487303Z","shell.execute_reply.started":"2023-01-02T17:03:41.487103Z","shell.execute_reply":"2023-01-02T17:03:41.487120Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Training Model:\n### In process, we will adding link at the final of the competition","metadata":{"execution":{"iopub.status.busy":"2023-01-02T17:03:41.489552Z","iopub.status.idle":"2023-01-02T17:03:41.490042Z","shell.execute_reply.started":"2023-01-02T17:03:41.489838Z","shell.execute_reply":"2023-01-02T17:03:41.489857Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<div class=\"alert alert-block alert-warning\" style=\"font-size:15px\">\nThank you for reading ‚ù§Ô∏èüöÄ\n</div>","metadata":{}},{"cell_type":"markdown","source":"# References","metadata":{}},{"cell_type":"markdown","source":"\n\nGilbert, W., Howard Hughes Medical Institute, Basic Sciences Division, Fred Hutchinson Cancer Research Center & Seattle. (1992, noviembre). Amino acid substitution matrices from protein blocks. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC50453/pdf/pnas01096-0363.pdf\n\nFuseSchool - Global Education. (2017, July 19). Enzymes | Cells | Biology | FuseSchool [Video]. YouTube. https://www.youtube.com/watch?v=rlH1ym916Fo\n\nNovozymes Enzyme Stability Prediction | Kaggle. (n.d.). https://www.kaggle.com/competitions/novozymes-enzyme-stability-prediction\n\nBack To Back SWE. (2019, January 11). Edit Distance Between 2 Strings - The Levenshtein Distance (‚ÄúEdit Distance‚Äù on LeetCode) [Video]. YouTube. https://www.youtube.com/watch?v=MiqoA-yF-0M\n\n\nBack To Back SWE. (2019, January 11). Edit Distance Between 2 Strings - The Levenshtein Distance (‚ÄúEdit Distance‚Äù on LeetCode) [Video]. YouTube. https://www.youtube.com/watch?v=MiqoA-yF-0M\n\nRAPIDs cudf tutorial. (2020, June 18). Kaggle. https://www.kaggle.com/code/beniel/rapids-cudf-tutorial\n\nThe Editors of Encyclopaedia Britannica. (2022, November 9). PH | Definition, Uses, & Facts. Encyclopedia Britannica. https://www.britannica.com/science/pH\n\nEffect of Temperature on Enzymatic Reaction - Creative Enzymes. (n.d.). https://www.creative-enzymes.com/resource/effect-of-temperature-on-enzymatic-reaction_50.html\n\nAmino acid substitution models. (2016, July 3). Evolution and Genomics. https://evomics.org/resources/substitution-models/amino-acid-substitution-models/\n\nAligning multiple protein sequences | UniProt. (n.d.). https://www.ebi.ac.uk/training/online/courses/uniprot-exploring-protein-sequence-and-functional-info/how-to-use-uniprot-tools-clone/aligning-multiple-protein-sequences/\n\nwwPDB.org. (n.d.). wwPDB: File Format. https://www.wwpdb.org/documentation/file-format\n\nNovozymes Enzyme Stability Prediction | Kaggle. (n.d.-b). https://www.kaggle.com/competitions/novozymes-enzyme-stability-prediction/discussion/354476\n\nTrain Data Contains Mutations Like Test Data! (2022, September 26). Kaggle. https://www.kaggle.com/code/cdeotte/train-data-contains-mutations-like-test-data\n\nNESP: changes EDA and baseline. (2022, October 4). Kaggle. https://www.kaggle.com/code/lucasmorin/nesp-changes-eda-and-baseline","metadata":{}}]}